modules/core/04-channel/ante.go

```diff
@@ -0,0 +1,57 @@
+package channel
+
+import (
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/cosmos/ibc-go/modules/core/04-channel/keeper"
+	"github.com/cosmos/ibc-go/modules/core/04-channel/types"
+)
+
+type ChannelAnteDecorator struct {
+	k keeper.Keeper
+}
+
+func NewChannelAnteDecorator(k keeper.Keeper) ChannelAnteDecorator {
+	return ChannelAnteDecorator{k: k}
+}
+
+func (cad ChannelAnteDecorator) AnteHandle(ctx sdk.Context, tx sdk.Tx, simulate bool, next sdk.AnteHandler) (sdk.Context, error) {
+	// do not run redundancy check on DeliverTx or simulate
+	if (ctx.IsCheckTx() || ctx.IsReCheckTx()) && !simulate {
+		// keep track of total packet messages and number of redundancies across `RecvPacket`, `AcknowledgePacket`, and `TimeoutPacket/OnClose`
+		msgs := 0
+		redundancies := 0
+		for _, m := range tx.GetMsgs() {
+			switch msg := m.(type) {
+            case *types.MsgRecvPacket:
+                if _, found := cad.k.GetPacketReceipt(ctx, msg.Packet.GetDestPort(), msg.Packet.GetDestChannel(), msg.Packet.GetSequence()); found {
+                    redundancies += 1
+                }  
+
+            case *types.MsgAcknowledgement:
+                if commitment := cad.k.GetPacketCommitment(ctx, msg.Packet.GetSourcePort(), msg.Packet.GetSourceChannel(), msg.Packet.GetSequence()); len(commitment) == 0 {
+                    redundancies += 1
+                }  
+
+            case *types.MsgTimeout:
+                if commitment := cad.k.GetPacketCommitment(ctx, msg.Packet.GetSourcePort(), msg.Packet.GetSourceChannel(), msg.Packet.GetSequence()); len(commitment) == 0 {
+                    redundancies += 1
+                }  
+
+            case *types.MsgTimeoutOnClose:
+                if commitment := cad.k.GetPacketCommitment(ctx, msg.Packet.GetSourcePort(), msg.Packet.GetSourceChannel(), msg.Packet.GetSequence()); len(commitment) == 0 {
+                    redundancies += 1
+                }  
+            default:
```

Comment:
One more change could be made I think. If the transaction contains an non IBC update message or relay message, it should be processed as normal. Not sure what non IBC messages would be batched with relay messages but if the tx contains a misbehaviour message, we shouldn't discard it based on redundant relays

Reply:
Ehh, i'd prefer not to make things more complicated for a quickfix. I see no reason why anyone would do that. And outside of narrow usecases like redundant update/redundant receives; i think we should stick to the atomic model of multimsg txs

Reply:
Hmm, but this is a long term fix in ibc-go. If a relayer batches a `MsgTransfer` with redundant relay messages, the `MsgTransfer` would be discarded without the relayer knowing

Reply:
Hmm but then I'd have to not count UpdateClient, since if it's valid but everything else isn't then what's the point. So it just keeps introducing edge cases. I'd rather keep the antedecorator simple and have relayer implementations adapt to it. Also since it's on `CheckTx` and `RecheckTx`, it's trivial to change at a later time